server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/integrity/VerifyRepositoryIntegrityAction.java

```diff
@@ -0,0 +1,260 @@
+        private final List<RepositoryVerificationException> exceptions;
+
+        public Response(List<RepositoryVerificationException> exceptions) {
+            this.exceptions = exceptions;
+        }
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            this.exceptions = in.readList(RepositoryVerificationException::new);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeList(exceptions);
+        }
+
+        @Override
+        public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params outerParams) {
+            return Iterators.concat(
+                Iterators.single((builder, params) -> builder.startObject().startArray("errors")),
+                exceptions.stream().<ToXContent>map(e -> (builder, params) -> {
+                    builder.startObject();
+                    ElasticsearchException.generateFailureXContent(builder, params, e, true);
+                    return builder.endObject();
+                }).iterator(),
+                Iterators.single((builder, params) -> builder.endArray().endObject())
+            );
+        }
+
+        @Override
+        public RestStatus getRestStatus() {
+            if (exceptions.isEmpty()) {
+                return RestStatus.OK;
+            } else {
+                return RestStatus.INTERNAL_SERVER_ERROR;
+            }
+        }
+
+        public List<RepositoryVerificationException> getExceptions() {
+            return exceptions;
+        }
+    }
+
+    public static class TransportAction extends TransportMasterNodeReadAction<Request, Response> {
+
+        private final RepositoriesService repositoriesService;
+
+        @Inject
+        public TransportAction(
+            TransportService transportService,
+            ClusterService clusterService,
+            RepositoriesService repositoriesService,
+            ThreadPool threadPool,
+            ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver
+        ) {
+            super(
+                NAME,
+                true,
+                transportService,
+                clusterService,
+                threadPool,
+                actionFilters,
+                Request::new,
+                indexNameExpressionResolver,
+                Response::new,
+                ThreadPool.Names.SNAPSHOT_META
+            );
+            this.repositoriesService = repositoriesService;
+        }
+
+        @Override
+        protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+            return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+        }
+
+        @Override
+        protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
+            // TODO add mechanism to block blob deletions while this is running
```

Comment:
And yes you are right that this will yield false errors if blobs are deleted out from underneath us. I need advice from @original-brownbear about whether there's a nice way to prevent that. I mean we could make it a stop-the-world thing like repo cleanup but that seems overkill. In principle we should even be able to let snapshotting continue as long as we just held back any finalisation-time deletes. I just worry that this could be a bigger change to the state machine, which IMO isn't worth making any more complex just for this.

Reply:
Yeah, I also think modifying the state machine for this seems like the wrong choice.

For deletions, I wonder if we could recheck that any missing blobs are still referenced/expected after finding that they are missing? I am not entirely sure of the order, but I think we delete the reference to the file before deleting the file?

Reply:
The number of potential anomalies is huge, and not all of them relate to missing blobs, so it's not clear that we can go back through the list and drop all the spurious ones.

I've made it so that it checks (and reports) the repo generation at the end of the analysis, so that at least you can see that the repo changed underneath us and therefore that some anomalies may be expected. I think this plus some docs about it will be sufficient for now.

Reply:
:thu: sounds like a pragmatic solution.